[DM-27531] Add a token UI

rra opened the pull request:
Includes some backend modifications to make the token UI easier to write and
to fix some bugs noticed while writing it.

rra commented:
This is my first JavaScript, so please also review for design and code organization and whatnot and feel free to nit-pick as aggressively as you would like! I don't know what I'm doing and am still learning.

jonathansick reviewed (commented):
Here are some preliminary notes trying to get gafaelfawr and the UI up in my local environment:

- I had an unexpectedly hard time running gafaelfawr on a Mac. The issue is the OpenSSL isn't automatically in the linker's path, so I have to run `export LDFLAGS="-L/usr/local/opt/openssl/lib"` in all terminal sessions. In other projects, I use `psycopg2-binary`. It could be worth either switching to pre-built binaries or adding some notes for development.

- When doing `tox -e run` I got the error 

   ```
     File "/Users/jsick/lsst/gafaelfawr/src/gafaelfawr/config.py", line 618, in _load_secret
    with open(path, "rb") as fh:
   FileNotFoundError: [Errno 2] No such file or directory: 'examples/secrets/github-client-  secret'
   ```

I'll continue to review just from the code standpoint, but I just wanted to note that the development environment reproducibility and documentation could take a look.

jonathansick commented on .github/workflows/ci.yaml:
```diff
@@ -20,6 +20,25 @@ jobs:
           python-version: ${{ matrix.python }}
 
+      - name: Set up Node
+        uses: actions/setup-node@v1
```
In some projects I've had issues where packages are incompatible with a new version of node. So what I've tended to do is use [nvm](https://github.com/nvm-sh/nvm) to "pin" a node version in my local environment with a `.nvmrc` file. Then in GitHub Actions I use that same file to set the node version: https://github.com/lsst-sqre/www_lsst_io/blob/ead7d71212b56431288f752ed5547347f7ae9634/.github/workflows/ci.yaml#L17-L24

Not to say that the same problem will pop up here, especially if packages are being continuously kept up to date, but I think its a nice technique.

rra replied:
`nvm` isn't packaged for Debian, so it's a bit painful to try to use this overall development path, but it does seem like a good idea to do GitHub Actions testing with a consistent version of Node. I'll adopt this and pin to the version of Node in Debian, I think.

jonathansick reviewed (approved):
I don't have a tonne of React experience given that I've only shipped one app in product before, but I can say that this foundation should work out well. Styling it should be fine. I also learned a lot of things along the way about formik, react-tables, and react-icons, so that was great!

rra commented:
I looked at `psycopg2-binary`, but the `psycopg2` documentation specifically says to not use that for a production deployment, so I shied away from it.

> The binary package is a practical choice for development and testing but in production it is advised to use the package built from sources.

I wasn't sure how to use it only for development and testing and not for production. Sorry that it's a pain on macOS! I added information about that to the documentation.

For the local development server, I added instructions to the documentation. See if the instructions at:

https://gafaelfawr.lsst.io/v/DM-27531/dev/development.html#starting-a-development-server

work for you.

rra commented on ui/src/components/errorBanner.js:
```diff
@@ -0,0 +1,14 @@
+
+export default function ErrorBanner({ error, id }) {
+  return error ? (
+    <div id={id}>
```
The thought was to make these findable in Selenium tests but I didn't finish the work. I'm going to drop `id` for now, since I think the long-term solution will be to replace this code with react-alert.

rra commented on ui/src/components/tokenModal.js:
```diff
@@ -0,0 +1,75 @@
+  onExit,
+}) {
+  const getApplicationNode = () => {
+    document.getElementById('application');
```
Indeed, I'm not. I just tried this and it worked and assumed that it wasn't really an example, but I think you're right that it's just an example. I'll add a `<div>` wrapper at the top level with id `application`.

rra commented on ui/src/components/tokenTable.js:
```diff
@@ -0,0 +1,154 @@
+import { useTable } from 'react-table';
+import { FaEdit, FaTrash } from 'react-icons/fa';
+
+function formatTimestamp(timestamp, { past }) {
```
I was fighting with eslint, which really didn't like inline arrow functions that returned components. This seems to be a known eslint bug that they're working on fixing, so I went back to the component style and just disabled the relevant eslint complaints for the arrow functions.

rra commented on ui/src/components/tokenTable.js:
```diff
@@ -0,0 +1,154 @@
+  const date = fromUnixTime(timestamp);
+  const relative = formatDistanceToNow(date, { addSuffix: past });
+  const absolute = date.toISOString().replace(/\.0+Z$/, 'Z');
+  return <span title={absolute}>{relative}</span>;
```
Oh, I didn't even know that existed. Good call!

rra commented on ui/src/hooks/error.js:
```diff
@@ -0,0 +1,20 @@
+  const onError = useCallback(
+    (error) => {
+      if (error instanceof TypeError) {
+        setError('Unable to contact backend API');
```
Were you running Gatsby in development mode? If so, it apparently returns 200 to unknown URLs, which is probably the problem. I've reworked this code a bit to try to capture that error and add a prefix to make it clear it was an API call problem.

rra commented on ui/src/functions/api.js:
```diff
@@ -0,0 +1,67 @@
+}
+
+// Redirect to the /login route and request a return to the current page.
+function apiLoginRedirect() {
```
The case I was trying to handle here is when the user is successfully logged in and browsing the site, and then their authentication cookies expire. I'm not sure how that works with a user profile context; we currently only call the login route once. The above pattern seems equivalent to putting the authentication handling in the `LoginContext`, but then everything starts failing with errors instead of a redirect when the user's credentials expire.

jonathansick commented:
Thanks, I'll try it today.

jonathansick reviewed (approved):
Great, I was able to bring up the development environment and make tokens.

I added a couple of line comments.

In the long term it would be nice to think of a way to set it up so that your test client ID isn't part of a committed file so that you don't need to remember not to commit that particular line. Maybe a make command that copies "source" example files into a personalizable example config file set that can be modified but aren't committed in git?

rra commented:
I think in the long term I want to simplify the nested structure of the configuration file, stop loading things from files, and rely on environment variables. Once that's done, people can just set environment variables with the client ID and secret to launch the local dev environment. I've created DM-28210 to track this.
